File: dotnet-desktop-guide/winforms/migration/snippets/clipboard-dataobject-net10/net/csharp/BinaryFormatterSupport.cs
using System;
using System.Collections.Generic;
using System.Reflection.Metadata;
using System.Windows.Forms;

namespace ClipboardExamples
{
    public class BinaryFormatterSupport
    {
        [Serializable]
        public class Person
        {
            public string Name { get; set; }
            public int Age { get; set; }
        }

        [Serializable]
        public class AppSettings
        {
            public string Theme { get; set; }
            public bool AutoSave { get; set; }
        }

        [Serializable]
        public class MyCustomType
        {
            public string Data { get; set; }
        }

        // <SecureTypeResolver>
        // Create a security-focused type resolver
        private static Type SecureTypeResolver(TypeName typeName)
        {
            // Explicit allow-list of permitted types—add only what you need
            var allowedTypes = new Dictionary<string, Type>
            {
                [typeof(Person).FullName!]      = typeof(Person),
                [typeof(AppSettings).FullName!] = typeof(AppSettings),
                [typeof(string).FullName!]      = typeof(string),
                [typeof(int).FullName!]         = typeof(int),
                // Add only the specific types your application requires
            };

            // Only allow explicitly listed types - exact string match required
            if (allowedTypes.TryGetValue(typeName.FullName, out Type allowedType))
            {
                return allowedType;
            }

            // Reject any type not in the allow-list with clear error message
            throw new InvalidOperationException(
                $"Type '{typeName.FullName}' is not permitted for clipboard deserialization");
        }

        // Use the resolver with clipboard operations
        public static void UseSecureTypeResolver()
        {
            if (Clipboard.TryGetData("LegacyData", SecureTypeResolver, out MyCustomType data))
            {
                ProcessLegacyData(data);
            }
        }
        // </SecureTypeResolver>

        private static void ProcessLegacyData(MyCustomType data)
        {
            Console.WriteLine($"Processing legacy data: {data.Data}");
        }
    }
}

File: dotnet-desktop-guide/winforms/migration/snippets/clipboard-dataobject-net10/net/vb/BinaryFormatterSupport.vb
Imports System
Imports System.Collections.Generic
Imports System.Reflection.Metadata
Imports System.Windows.Forms

Namespace ClipboardExamples
    Public Class BinaryFormatterSupport
        <Serializable>
        Public Class Person
            Public Property Name As String
            Public Property Age As Integer
        End Class

        <Serializable>
        Public Class AppSettings
            Public Property Theme As String
            Public Property AutoSave As Boolean
        End Class

        <Serializable>
        Public Class MyCustomType
            Public Property Data As String
        End Class

        ' <SecureTypeResolver>
        ' Create a security-focused type resolver
        Private Shared Function SecureTypeResolver(typeName As TypeName) As Type
            ' Explicit allow-list of permitted types—add only what you need
            Dim allowedTypes As New Dictionary(Of String, Type) From {
                {GetType(Person).FullName, GetType(Person)},
                {GetType(AppSettings).FullName, GetType(AppSettings)},
                {GetType(String).FullName, GetType(String)},
                {GetType(Integer).FullName, GetType(Integer)}
            }
            ' Add only the specific types your application requires

            ' Only allow explicitly listed types - exact string match required
            Dim allowedType As Type = Nothing
            If allowedTypes.TryGetValue(typeName.FullName, allowedType) Then
                Return allowedType
            End If

            ' Reject any type not in the allow-list with clear error message
            Throw New InvalidOperationException(
                $"Type '{typeName.FullName}' is not permitted for clipboard deserialization")
        End Function

        ' Use the resolver with clipboard operations
        Public Shared Sub UseSecureTypeResolver()
            Dim data As MyCustomType = Nothing
            If Clipboard.TryGetData("LegacyData", AddressOf SecureTypeResolver, data) Then
                Console.WriteLine($"Processing legacy data: {data.Data}")
            End If
        End Sub
        ' </SecureTypeResolver>
    End Class
End Namespace
